acq4/modules/TaskRunner/analysisModules/Imaging/Imaging.py

```diff
@@ -37,13 +40,26 @@ def __init__(self, *args):
             dict(name='downsample', type='int', value=1, suffix='x', bounds=[1,None]),
             dict(name='display', type='bool', value=True),
             dict(name='scanProgram', type='list', values=[]),
+            dict(name='Objective', type='str', value='Unknown', readonly=True),
+            dict(name='Filter', type='str', value='Unknown', readonly=True),
             ])
         self.ptree.setParameters(self.params, showTop=False)
         self.params.sigTreeStateChanged.connect(self.update)
         self.params.child('detectors').sigAddNew.connect(self.addDetectorClicked)
         self.params.child('decomb', 'auto').sigActivated.connect(self.autoDecomb)
 
         self.man = getManager()
+        print self.man, dir(self.man.devices)
+        self.scannerDev = self.man.getDevice('Scanner')
+        # find first scope device that is parent of scanner
+        dev = self.scannerDev
+        while dev is not None and not isinstance(dev, Microscope):
+            dev = dev.parentDevice()
+        self.scopeDev = dev
+        self.filterwheel = self.man.getDevice('FilterWheel')
```

Comment:
Hi Luke,

thank you for all the input. I implemented everything except for your last
suggestion ... have to run now and will continue later.

Cheers,
Mike

On Sun, May 8, 2016 at 7:58 AM, Luke Campagnola <notifications@github.com>
wrote:

> In acq4/modules/TaskRunner/analysisModules/Imaging/Imaging.py
> <https://github.com/acq4/acq4/pull/45#discussion_r62427303>:
>
> >              ])
> >          self.ptree.setParameters(self.params, showTop=False)
> >          self.params.sigTreeStateChanged.connect(self.update)
> >          self.params.child('detectors').sigAddNew.connect(self.addDetectorClicked)
> >          self.params.child('decomb', 'auto').sigActivated.connect(self.autoDecomb)
> >
> >          self.man = getManager()
> > +        print self.man, dir(self.man.devices)
> > +        self.scannerDev = self.man.getDevice('Scanner')
> > +        # find first scope device that is parent of scanner
> > +        dev = self.scannerDev
> > +        while dev is not None and not isinstance(dev, Microscope):
> > +            dev = dev.parentDevice()
> > +        self.scopeDev = dev
> > +        self.filterwheel = self.man.getDevice('FilterWheel')
>
> Obviously this won't work on systems without a device with this name.. I
> think the correct approach here would be something like: 1) make the PMT an
> optomech device and set up the hierarchy like Stage -> Microscope ->
> FilterWheel -> PMT; 2) ask the PMT whether it has a filter selection in its
> parent devices. To start, you could do this pretty simply:
>
> class PhotoDetector(DAQGeneric, OptomechDevice):
>     def __init__(self, dm, config, name):
>         omConf = {}
>         for k in ['parentDevice', 'transform']:
>             if k in config:
>                 omConf[k] = config.pop(k)
>         OptomechDevice.__init__(self, dm, config, name)
>         DAQGeneric.__init__(self, dm, config, name)
>
>     def getFilterDevice(self):
>         # return parent filter device or None
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/acq4/acq4/pull/45/files/5a99fb27df27f5ded91a35f8bb9542993c512823#r62427303>
>



-- 
Michael Graupner (PhD)
Brain Physiology Lab, Université Paris Descartes
http://www.biomedicale.parisdescartes.fr/~mgraupe/
michael.graupner@parisdescartes.fr